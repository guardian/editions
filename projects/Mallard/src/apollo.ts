/**
 * Setup Apollo framework. At the moment we use Apollo Local State management
 * mechanism and don't fetch data from GraphQL endpoints.
 */
import { ApolloClient } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { SETTINGS_RESOLVERS } from './helpers/settings/resolvers'
import { resolveWeather } from './helpers/location'

/**
 * Resolvers is what Apollo uses to get the value of field that has never been
 * cached before. For example if you have a query `{ foo @client }`, Apollo
 * looks for the field `foo` at the root of the cache. If not available, it'll
 * look for the resolver `Query.foo` (`Query` is just the "namespace" where
 * all queries start).
 *  *
 * Resolvers are functions, which can be asynchonous. They can gather data from
 * local storage (ex. settings), from local APIs (ex. permissions, geolocation),
 * from remote APIs (ex. REST API).
 *
 * Once a resolver returns a value, Apollo will refresh automatically
 * any component that was requiring this in a query via the `useQuery` function.
 */
const RESOLVERS = {
    Query: {
        ...SETTINGS_RESOLVERS,
        weather: resolveWeather,
    },
}

/**
 * The `Link` is the interface Apollo uses to reach GraphQL servers. Since we
 * don't fetch anything from GraphQL endpoints at the moment, we use a no-op
 * `Link` object. We *must* throw to prevent bad queries from carrying through.
 *
 * Apollo would try to use the `Link` if you have a query such as `{ foo }`
 * without the `@client` annotation. In our current use case, `@client` always
 * needs to be specified.
 */
const link = {
    split: () => {
        throw new Error('Not implemented. Did you forget to use `@client`?')
    },
    concat: () => {
        throw new Error('not implemented Did you forget to use `@client`?')
    },
    request: () => {
        throw new Error('not implemented Did you forget to use `@client`?')
    },
}

export const createApolloClient = () =>
    new ApolloClient({
        cache: new InMemoryCache(),
        link,
        resolvers: RESOLVERS,
    })
