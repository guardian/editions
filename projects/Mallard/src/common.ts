export * from '../../Apps/common/src/index';
