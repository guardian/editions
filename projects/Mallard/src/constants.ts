import { Platform } from 'react-native';
import Config from 'react-native-config';

const {
	ID_API_URL,
	MEMBERS_DATA_API_URL,
	ID_ACCESS_TOKEN,
	ITUNES_CONNECT_SHARED_SECRET,
	ANDROID_RELEASE_STREAM,
} = Config;

const FACEBOOK_CLIENT_ID = '528503751025697';

const AUTH_TTL = 86400000; // ms in a day

const CAS_ENDPOINT_URL = 'https://content-auth.guardian.co.uk/subs';

const GOOGLE_CLIENT_ID =
	Platform.OS === 'android'
		? __DEV__
			? '774465807556-5oq58e8pnfqn9sptsivri5kcs9mthef9'
			: '774465807556-cjat38acttpl7md4nfc4pfediouh7v97'
		: '774465807556-kgaj5an4pc4fmr3svp5nfpulekc1rl3n';

const LEGACY_SUBSCRIBER_ID_USER_DEFAULT_KEY = 'printSubscriberID';
const LEGACY_SUBSCRIBER_POSTCODE_USER_DEFAULT_KEY = 'printSubscriberPostcode';

const JOIN_BETA_LINK = Platform.select({
	ios: 'https://testflight.apple.com/join/O2EojUEl',
	android: 'https://play.google.com/apps/testing/com.guardian.editions',
});

export {
	CAS_ENDPOINT_URL,
	ID_API_URL,
	MEMBERS_DATA_API_URL,
	ID_ACCESS_TOKEN,
	FACEBOOK_CLIENT_ID,
	GOOGLE_CLIENT_ID,
	AUTH_TTL,
	LEGACY_SUBSCRIBER_ID_USER_DEFAULT_KEY,
	LEGACY_SUBSCRIBER_POSTCODE_USER_DEFAULT_KEY,
	ITUNES_CONNECT_SHARED_SECRET,
	ANDROID_RELEASE_STREAM,
	JOIN_BETA_LINK,
};
