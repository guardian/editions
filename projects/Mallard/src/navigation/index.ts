import { useEffect } from 'react'
import {
    createAppContainer,
    createStackNavigator,
    createSwitchNavigator,
    NavigationScreenProp,
    StackViewTransitionConfigs,
    NavigationTransitionProps,
} from 'react-navigation'
import { AuthSwitcherScreen } from 'src/screens/identity-login-screen'
import { OnboardingConsentScreen } from 'src/screens/onboarding-screen'
import { AlreadySubscribedScreen } from 'src/screens/settings/already-subscribed-screen'
import { SubscriptionDetailsScreen } from 'src/screens/settings/subscription-details-screen'
import { ApiScreen } from 'src/screens/settings/api-screen'
import { CasSignInScreen } from 'src/screens/settings/cas-sign-in-screen'
import { CreditsScreen } from 'src/screens/settings/credits-screen'
import { FAQScreen } from 'src/screens/settings/faq-screen'
import {
    GdprConsentScreen,
    GdprConsentScreenForOnboarding,
} from 'src/screens/settings/gdpr-consent-screen'
import { HelpScreen } from 'src/screens/settings/help-screen'
import {
    PrivacyPolicyScreen,
    PrivacyPolicyScreenForOnboarding,
} from 'src/screens/settings/privacy-policy-screen'
import { TermsAndConditionsScreen } from 'src/screens/settings/terms-and-conditions-screen'
import { color } from 'src/theme/color'
import { ArticleScreen } from '../screens/article-screen'
import { HomeScreen } from '../screens/home-screen'
import { IssueScreen } from '../screens/issue-screen'
import { SettingsScreen } from '../screens/settings-screen'
import { mapNavigationToProps } from './helpers/base'
import { createArticleNavigator } from './navigators/article'
import { createHeaderStackNavigator } from './navigators/header'
import { createModalNavigator } from './navigators/modal'
import { createUnderlayNavigator } from './navigators/underlay'
import { routeNames } from './routes'
import { useQuery } from 'src/hooks/apollo'
import gql from 'graphql-tag'
import { ManageEditionsScreen } from 'src/screens/settings/manage-editions-screen'
import { WeatherGeolocationConsentScreen } from 'src/screens/weather-geolocation-consent-screen'

const navOptionsWithGraunHeader = {
    headerStyle: {
        backgroundColor: color.primary,
        borderBottomColor: color.text,
    },
    headerTintColor: color.textOverPrimary,
}

/* The screens you add to IOS_MODAL_ROUTES will have the modal transition. I.e they will slide up from the bottom. */
const IOS_MODAL_ROUTES = [
    routeNames.onboarding.PrivacyPolicyInline,
    routeNames.onboarding.OnboardingConsentInline,
]

const dynamicModalTransition = (
    transitionProps: NavigationTransitionProps,
    prevTransitionProps: NavigationTransitionProps,
) => {
    const isModal = IOS_MODAL_ROUTES.some(
        screenName =>
            screenName === transitionProps.scene.route.routeName ||
            (prevTransitionProps &&
                screenName === prevTransitionProps.scene.route.routeName),
    )

    return StackViewTransitionConfigs.defaultTransitionConfig(
        transitionProps,
        prevTransitionProps,
        isModal,
    )
}

const AppStack = createModalNavigator(
    createUnderlayNavigator(
        createArticleNavigator(IssueScreen, ArticleScreen),
        {
            [routeNames.IssueList]: HomeScreen,
        },
    ),
    {
        [routeNames.ManageEditions]: createHeaderStackNavigator({
            [routeNames.ManageEditions]: ManageEditionsScreen,
        }),
        [routeNames.Settings]: createHeaderStackNavigator(
            {
                [routeNames.Settings]: SettingsScreen,
                [routeNames.Endpoints]: ApiScreen,
                [routeNames.GdprConsent]: GdprConsentScreen,
                [routeNames.PrivacyPolicy]: PrivacyPolicyScreen,
                [routeNames.TermsAndConditions]: TermsAndConditionsScreen,
                [routeNames.Help]: HelpScreen,
                [routeNames.Credits]: CreditsScreen,
                [routeNames.FAQ]: FAQScreen,
                [routeNames.AlreadySubscribed]: AlreadySubscribedScreen,
                [routeNames.SubscriptionDetails]: SubscriptionDetailsScreen,
            },
            {
                defaultNavigationOptions: {
                    ...navOptionsWithGraunHeader,
                },
            },
        ),
        [routeNames.WeatherGeolocationConsent]: createHeaderStackNavigator({
            [routeNames.WeatherGeolocationConsent]: WeatherGeolocationConsentScreen,
        }),
    },
)

const OnboardingStack = createModalNavigator(
    createStackNavigator(
        {
            [routeNames.onboarding.OnboardingConsent]: createStackNavigator(
                {
                    Main: {
                        screen: mapNavigationToProps(
                            OnboardingConsentScreen,
                            nav => ({
                                onContinue: () => nav.navigate('App'),
                                onOpenGdprConsent: () =>
                                    nav.navigate(
                                        routeNames.onboarding
                                            .OnboardingConsentInline,
                                    ),
                                onOpenPrivacyPolicy: () =>
                                    nav.navigate(
                                        routeNames.onboarding
                                            .PrivacyPolicyInline,
                                    ),
                            }),
                        ),
                        navigationOptions: {
                            header: null,
                        },
                    },
                    [routeNames.onboarding
                        .OnboardingConsentInline]: GdprConsentScreenForOnboarding,
                    [routeNames.onboarding
                        .PrivacyPolicyInline]: PrivacyPolicyScreenForOnboarding,
                },
                {
                    headerMode: 'none',
                    transitionConfig: dynamicModalTransition,
                    defaultNavigationOptions: {
                        ...navOptionsWithGraunHeader,
                    },
                },
            ),
        },
        {
            headerMode: 'none',
        },
    ),
    {},
)

const ONBOARDING_QUERY = gql(`{
    gdprAllowEssential @client
    gdprAllowPerformance @client
    gdprAllowFunctionality @client
}`)

type OnboardingQueryData = {
    gdprAllowEssential: boolean
    gdprAllowPerformance: boolean
    gdprAllowFunctionality: boolean
}

const hasOnboarded = (data: OnboardingQueryData) => {
    return (
        data.gdprAllowEssential != null &&
        data.gdprAllowFunctionality != null &&
        data.gdprAllowPerformance != null
    )
}

const RootNavigator = createAppContainer(
    createStackNavigator(
        {
            AppRoot: createSwitchNavigator(
                {
                    Main: ({
                        navigation,
                    }: {
                        navigation: NavigationScreenProp<{}>
                    }) => {
                        const query = useQuery<OnboardingQueryData>(
                            ONBOARDING_QUERY,
                        )
                        useEffect(() => {
                            /** Setting is still loading, do nothing yet. */
                            if (query.loading) return
                            const { data } = query
                            // If any flag is unknown still, we want to onboard.
                            // We expected people to give explicit yay/nay to
                            // each GDPR bucket.
                            if (!hasOnboarded(data)) {
                                navigation.navigate('Onboarding')
                            } else {
                                navigation.navigate('App')
                            }
                        })
                        return null
                    },
                    App: AppStack,
                    Onboarding: OnboardingStack,
                },
                {
                    initialRouteName: 'Main',
                },
            ),
            [routeNames.SignIn]: AuthSwitcherScreen,
            [routeNames.CasSignIn]: CasSignInScreen,
        },
        {
            initialRouteName: 'AppRoot',
            mode: 'modal',
            headerMode: 'none',
        },
    ),
)

export { RootNavigator }
