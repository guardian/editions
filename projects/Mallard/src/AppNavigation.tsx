import {
	CardStyleInterpolators,
	createStackNavigator,
} from '@react-navigation/stack';
import React from 'react';
import { Animated } from 'react-native';
import type { RootStackParamList } from './navigation/NavigationModels';
import { RouteNames } from './navigation/NavigationModels';
import { ArticleWrapper } from './navigation/navigators/article';
import { EditionsMenuScreen } from './screens/editions-menu-screen';
import { HomeScreen } from './screens/home-screen';
import { AuthSwitcherScreen } from './screens/identity-login-screen';
import { IssueScreen } from './screens/issue-screen';
import { OnboardingConsentScreen } from './screens/onboarding-screen';
import { SettingsScreen } from './screens/settings-screen';
import { AlreadySubscribedScreen } from './screens/settings/already-subscribed-screen';
import { ApiScreen } from './screens/settings/api-screen';
import { BetaProgrammeFAQsScreen } from './screens/settings/beta-programme-faqs';
import { CasSignInScreen } from './screens/settings/cas-sign-in-screen';
import { CreditsScreen } from './screens/settings/credits-screen';
import { EditionsScreen } from './screens/settings/editions-screen';
import { FAQScreen } from './screens/settings/faq-screen';
import {
	GdprConsentScreen,
	GdprConsentScreenForOnboarding,
} from './screens/settings/gdpr-consent-screen';
import { HelpScreen } from './screens/settings/help-screen';
import { ManageEditionsScreenWithHeader } from './screens/settings/manage-editions-screen';
import {
	PrivacyPolicyScreen,
	PrivacyPolicyScreenForOnboarding,
} from './screens/settings/privacy-policy-screen';
import { SubscriptionDetailsScreen } from './screens/settings/subscription-details-screen';
import { TermsAndConditionsScreen } from './screens/settings/terms-and-conditions-screen';

const Stack = createStackNavigator<RootStackParamList>();

const { multiply } = Animated;

const cardStyleInterpolator = (props: any) => {
	const translateX = multiply(
		props.current.progress.interpolate({
			inputRange: [0, 1],
			outputRange: [props.layouts.screen.width, 0],
			extrapolate: 'clamp',
		}),
		props.inverted,
	);

	return {
		// ...CardStyleInterpolators.forHorizontalIOS(props),
		cardStyle: {
			overflow: 'hidden',
			transform: [
				// Translation for the animation of the current card
				{
					translateX,
				},
			],
		},
		overlayStyle: {
			opacity: props.current.progress.interpolate({
				inputRange: [0, 1],
				outputRange: [0, 0.5],
				extrapolate: 'clamp',
			}),
		},
	};
};

const RootStack = () => {
	return (
		<Stack.Navigator
			initialRouteName={RouteNames.Home}
			screenOptions={{ gestureEnabled: false, headerShown: false }}
		>
			<Stack.Screen
				name={RouteNames.Issue}
				component={IssueScreen}
				options={{}}
			/>
			<Stack.Screen
				name={RouteNames.IssueList}
				component={HomeScreen}
				options={{
					gestureDirection: 'horizontal',
					cardStyle: { backgroundColor: 'transparent' },
					cardOverlayEnabled: true,
					cardStyleInterpolator,
				}}
			/>
			<Stack.Screen
				name={RouteNames.EditionsMenu}
				component={EditionsMenuScreen}
				options={{
					gestureDirection: 'horizontal-inverted',
					cardStyle: { backgroundColor: 'transparent' },
					cardOverlayEnabled: true,
					cardStyleInterpolator,
				}}
			/>
			<Stack.Screen
				name={RouteNames.Article}
				component={ArticleWrapper}
				options={{
					cardStyleInterpolator:
						CardStyleInterpolators.forModalPresentationIOS,
					gestureEnabled: true,
					gestureDirection: 'vertical',
				}}
			/>
			{/* <Stack.Screen
                name={RouteNames.Article}
                component={SlideCardJames}
                options={{
                    cardStyleInterpolator: props => {
                        return {
                            ...CardStyleInterpolators.forModalPresentationIOS(
                                props,
                            ),
                        }
                    },
                }}
            /> */}
			<Stack.Screen
				name={RouteNames.Settings}
				component={SettingsScreen}
			/>
			<Stack.Screen
				name={RouteNames.SignIn}
				component={AuthSwitcherScreen}
			/>
			{/* <Stack.Screen
                name={RouteNames.OnboardingStart}
            /> */}
			<Stack.Screen
				name={RouteNames.OnboardingConsent}
				component={OnboardingConsentScreen}
			/>
			<Stack.Screen
				name={RouteNames.OnboardingConsentInline}
				component={GdprConsentScreenForOnboarding}
			/>
			<Stack.Screen
				name={RouteNames.PrivacyPolicyInline}
				component={PrivacyPolicyScreenForOnboarding}
			/>
			<Stack.Screen
				name={RouteNames.AlreadySubscribed}
				component={AlreadySubscribedScreen}
			/>
			<Stack.Screen
				name={RouteNames.CasSignIn}
				component={CasSignInScreen}
			/>
			<Stack.Screen
				name={RouteNames.ManageEditionsSettings}
				component={ManageEditionsScreenWithHeader}
			/>
			{/** @TODO Fix the enable all button */}
			<Stack.Screen
				name={RouteNames.GdprConsent}
				component={GdprConsentScreen}
			/>
			<Stack.Screen
				name={RouteNames.PrivacyPolicy}
				component={PrivacyPolicyScreen}
			/>

			{/* ==== Inspect from here === */}
			<Stack.Screen
				name={RouteNames.Edition}
				component={EditionsScreen}
			/>
			<Stack.Screen
				name={RouteNames.TermsAndConditions}
				component={TermsAndConditionsScreen}
			/>
			<Stack.Screen
				name={RouteNames.BetaProgrammeFAQs}
				component={BetaProgrammeFAQsScreen}
			/>
			<Stack.Screen name={RouteNames.Help} component={HelpScreen} />
			<Stack.Screen name={RouteNames.Credits} component={CreditsScreen} />
			<Stack.Screen name={RouteNames.FAQ} component={FAQScreen} />
			<Stack.Screen
				name={RouteNames.SubscriptionDetails}
				component={SubscriptionDetailsScreen}
			/>
			<Stack.Screen name={RouteNames.Endpoints} component={ApiScreen} />
			{/* Turned off to remove Promise rejection error on Android */}
			{/* <Stack.Screen
                name={RouteNames.Storybook}
                component={StorybookScreen}
            /> */}
		</Stack.Navigator>
	);
};

export { RootStack };
