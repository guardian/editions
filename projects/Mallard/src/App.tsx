// Welcome to the main entry point of the app.
//
// In this file, we'll be kicking off our app or storybook.

import { ApolloProvider } from '@apollo/react-hooks';
import AsyncStorage from '@react-native-community/async-storage';
import type ApolloClient from 'apollo-client';
import React from 'react';
import { AppState, StatusBar, StyleSheet, View } from 'react-native';
import SplashScreen from 'react-native-splash-screen';
import { eventEmitter } from 'src/helpers/event-emitter';
import { weatherHider } from 'src/helpers/weather-hider';
import {
	ConfigProvider,
	largeDeviceMemory,
} from 'src/hooks/use-config-provider';
import { NavPositionProvider } from 'src/hooks/use-nav-position';
import { setUserId } from 'src/services/ophan';
import { AppNavigation } from './AppNavigation';
import { AccessProvider } from './authentication/AccessContext';
import type { IdentityAuthData } from './authentication/authorizers/IdentityAuthorizer';
import type { AnyAttempt } from './authentication/lib/Attempt';
import { isValid } from './authentication/lib/Attempt';
import { BugButtonHandler } from './components/Button/BugButtonHandler';
import { ErrorBoundary } from './components/layout/ui/errors/error-boundary';
import { Modal, ModalRenderer } from './components/modal';
import { NetInfoDevOverlay } from './components/NetInfoDevOverlay';
import { NetInfoAutoToast } from './components/toast/net-info-auto-toast';
import { prepareAndDownloadTodaysIssue } from './download-edition/prepare-and-download-issue';
import { prepFileSystem } from './helpers/files';
import { nestProviders } from './helpers/provider';
import { pushDownloadFailsafe } from './helpers/push-download-failsafe';
import { isInBeta } from './helpers/release-stream';
import { newMobileProdStack } from './helpers/settings/defaults';
import { setApiUrl } from './helpers/settings/setters';
import { rollbackedApiUrlForBetaUsers } from './helpers/storage';
import { EditionProvider } from './hooks/use-edition-provider';
import { SettingsOverlayProvider } from './hooks/use-settings-overlay';
import { ToastProvider } from './hooks/use-toast';
import { pushNotificationRegistration } from './notifications/push-notifications';
import { DeprecateVersionModal } from './screens/deprecate-screen';
import { apolloClient } from './services/apollo-singleton';
import { errorService } from './services/errors';
import { loggingService } from './services/logging';
import { remoteConfigService } from './services/remote-config';

// Log Intitialisation
if (!__DEV__) {
	errorService.init(apolloClient);
}
loggingService.init(apolloClient);
remoteConfigService.init();

// --- SETUP OPERATIONS ---
pushNotificationRegistration();
prepFileSystem();

const styles = StyleSheet.create({
	appContainer: {
		flex: 1,
		backgroundColor: '#000',
	},
});

const persistenceKey = 'dev-nav-key-232asfdffgdfg1asdffgfdgfdga3413';

const isReactNavPersistenceError = (e: Error) =>
	__DEV__ && e.message.includes('There is no route defined for');

const WithProviders = nestProviders(
	Modal,
	ToastProvider,
	NavPositionProvider,
	ConfigProvider,
	EditionProvider,
	SettingsOverlayProvider,
);

const handleIdStatus = (attempt: AnyAttempt<IdentityAuthData>) =>
	setUserId(isValid(attempt) ? attempt.data.userDetails.id : null);

const shouldHavePushFailsafe = async (client: ApolloClient<object>) => {
	const largeRAM = await largeDeviceMemory();
	if (largeRAM) {
		pushDownloadFailsafe(client);
	}
};

const forceUpdateApiUrlIfBeta = async () => {
	// TODO - Remove this whole function before release to prod.
	// Previously we forced beta users to use a tmp url to point to the new stack, now that
	// we changed the fastly config we can rollback urls and eventually we can remove this full method.
	if (isInBeta()) {
		const rollbackedApi = await rollbackedApiUrlForBetaUsers.get();
		if (!rollbackedApi) {
			setApiUrl(apolloClient, newMobileProdStack);
			console.log('*** Beta: rolling back to default prod api url ***');
			await rollbackedApiUrlForBetaUsers.set(true);
		} else {
			console.log('*** Beta: api url for Beta users already updated ***');
		}
	}
};

export default class App extends React.Component {
	componentDidMount() {
		SplashScreen.hide();
		prepareAndDownloadTodaysIssue(apolloClient);
		shouldHavePushFailsafe(apolloClient);
		loggingService.postLogs();

		AppState.addEventListener('change', async (appState) => {
			if (appState === 'active') {
				prepareAndDownloadTodaysIssue(apolloClient);
				loggingService.postLogs();
				forceUpdateApiUrlIfBeta();
			}
		});

		{
			eventEmitter.on('editionCachesSet', () => {
				weatherHider(apolloClient);
			});
		}
	}

	async componentDidCatch(e: Error) {
		/**
		 * use an heuristic to check whether this is a react-nav error
		 * if it is then ditch our persistence and try to re-render
		 */
		if (isReactNavPersistenceError(e)) {
			await AsyncStorage.removeItem(persistenceKey);
			this.forceUpdate();
		}
	}
	/**
	 * When the component is mounted. This happens asynchronously and simply
	 * re-renders when we're good to go.
	 */
	render() {
		return (
			<ErrorBoundary>
				<ApolloProvider client={apolloClient}>
					<NetInfoDevOverlay>
						<WithProviders>
							<AccessProvider
								onIdentityStatusChange={handleIdStatus}
							>
								<StatusBar
									barStyle="light-content"
									backgroundColor="#041f4a"
								/>
								<View style={styles.appContainer}>
									<AppNavigation />
									<NetInfoAutoToast />
								</View>
								<ModalRenderer />
								<BugButtonHandler />
								<DeprecateVersionModal />
							</AccessProvider>
						</WithProviders>
					</NetInfoDevOverlay>
				</ApolloProvider>
			</ErrorBoundary>
		);
	}
}

/**
 * This needs to match what's found in your app_delegate.m and MainActivity.java.
 */

export const APP_NAME = 'Mallard';
