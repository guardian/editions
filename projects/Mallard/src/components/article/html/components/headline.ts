import { html } from 'src/helpers/webview'
import { HeaderType, ArticleType, ArticlePillar } from 'src/common'
import { ArticleHeaderProps } from './header'
import { Quotes } from './icon/quotes'

const getHeadline = (
    articleHeaderType: HeaderType,
    articleType: ArticleType,
    headerProps: ArticleHeaderProps,
    pillar?: ArticlePillar,
) => {
    if (
        articleHeaderType === HeaderType.LargeByline ||
        articleType === ArticleType.Interview
    ) {
        return html`
            <h1>
                <span class="header-top-headline"
                    >${(articleType === ArticleType.Opinion ||
                        articleType === ArticleType.Interview) &&
                        Quotes()}${headerProps.headline}
                </span>
                ${articleType !== ArticleType.Interview &&
                    html`
                        <span class="header-top-byline"
                            >${headerProps.bylineHtml}
                        </span>
                    `}
            </h1>
        `
    } else {
        const pillarInclusion =
            pillar === 'news' || pillar === 'neutral' || pillar === 'opinion'
        console.log(pillarInclusion)
        const showcaseStyle =
            articleType === ArticleType.Showcase && pillarInclusion
                ? ` class="alt"`
                : ``
        return html`
            <h1${showcaseStyle}>
                ${headerProps.headline}
            </h1>
        `
    }
}

export { getHeadline }
