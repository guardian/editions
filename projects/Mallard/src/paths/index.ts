import RNFS from 'react-native-fs';
import type {
	CAPIArticle,
	Collection,
	Front,
	Image,
	ImageUse,
	Issue,
} from 'src/common';
import { frontPath, issuePath } from 'src/common';
import { imagePath } from '../../../Apps/common/src';
import { defaultRegionalEditions } from '../../../Apps/common/src/editions-defaults';

export interface PathToIssue {
	localIssueId: Issue['localId'];
	publishedIssueId: Issue['publishedId'];
}

export interface PathToArticle {
	collection: Collection['key'];
	front: Front['key'];
	article: CAPIArticle['key'];
	localIssueId: Issue['localId'];
	publishedIssueId: Issue['publishedId'];
}

export const APIPaths = {
	issue: issuePath,
	front: frontPath,
	image: imagePath,
};

const issuesDir = `${RNFS.DocumentDirectoryPath}/issues`;

const issueRoot = (localIssueId: string) => `${issuesDir}/${localIssueId}`;
const mediaRoot = (localIssueId: string) => `${issueRoot(localIssueId)}/media`;
const editionDir = (editionSlug: string) => {
	return `${issuesDir}/${editionSlug}`;
};
const edtionsDirList = async (): Promise<string[]> => {
	return defaultRegionalEditions.map((reg) => {
		return `${issuesDir}/${reg.edition}`;
	});
};

export const FSPaths = {
	issuesDir,
	editionDir,
	edtionsDirList,
	issueRoot,
	mediaRoot,
	image: (localIssueId: string, image: Image, use: ImageUse) =>
		imagePath(issueRoot(localIssueId), image, use),
	zip: (localIssueId: string, filename: string) =>
		`${issueRoot(localIssueId)}/${filename}.zip`,
	issue: (localIssueId: string) => `${issueRoot(localIssueId)}/issue`,
	front: (localIssueId: string, frontId: string) =>
		`${issueRoot(localIssueId)}/front/${frontId}`,
	downloadRoot: `${issuesDir}/download`,
	downloadIssueLocation: (localIssueId: string) =>
		`${issuesDir}/download/${localIssueId}`,
};
