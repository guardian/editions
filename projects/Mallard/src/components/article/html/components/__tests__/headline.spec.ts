import { getHeadline } from '../headline'
import { HeaderType, ArticleType } from 'src/common'

jest.mock('src/components/article/html/components/icon/quotes', () => ({
    Quotes: () => '<Quotes />',
}))

describe('getHeadline', () => {
    it('should display a headline with the byline and quotes', () => {
        const html = getHeadline(HeaderType.LargeByline, ArticleType.Opinion, {
            headline: 'Test Headline',
            bylineHtml: '<p>Test Byline</p>',
        })
        expect(html).toMatchSnapshot()
    })
    it('should display a headline with only the byline', () => {
        const html = getHeadline(HeaderType.LargeByline, ArticleType.Article, {
            headline: 'Test Headline',
            bylineHtml: '<p>Test Byline</p>',
        })
        expect(html).toMatchSnapshot()
    })
    it('should display a headline on its own', () => {
        const html = getHeadline(HeaderType.NoByline, ArticleType.Opinion, {
            headline: 'Test Headline',
            bylineHtml: '<p>Test Byline</p>',
        })
        expect(html).toMatchSnapshot()
    })
})
