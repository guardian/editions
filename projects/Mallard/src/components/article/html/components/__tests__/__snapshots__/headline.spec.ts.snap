// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getHeadline should display a headline on its own 1`] = `
"
            <h1>
                Test Headline
            </h1>
        "
`;

exports[`getHeadline should display a headline with only the byline 1`] = `
"
            <h1>
                
                <span class=\\"header-top-headline\\"
                    >Test Headline
                </span>
                <span class=\\"header-top-byline\\"
                    ><p>Test Byline</p>
                </span>
            </h1>
        "
`;

exports[`getHeadline should display a headline with the byline and quotes 1`] = `
"
            <h1>
                <Quotes />
                <span class=\\"header-top-headline\\"
                    >Test Headline
                </span>
                <span class=\\"header-top-byline\\"
                    ><p>Test Byline</p>
                </span>
            </h1>
        "
`;
